test/src/kernels/SimpleTestShapeElementKernel.C

```diff
@@ -0,0 +1,49 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+#include "SimpleTestShapeElementKernel.h"
+
+template<>
+InputParameters validParams<SimpleTestShapeElementKernel>()
+{
+  InputParameters params = validParams<Kernel>();
+  params.addRequiredParam<UserObjectName>("user_object", "Name of a SimpleTestShapeElementUserObject");
+  return params;
+}
+
+
+SimpleTestShapeElementKernel::SimpleTestShapeElementKernel(const InputParameters & parameters) :
+    Kernel(parameters),
+    _shp(getUserObject<SimpleTestShapeElementUserObject>("user_object")),
+    _shp_integral(_shp.getIntegral()),
+    _shp_jacobian(_shp.getJacobian())
+{
+}
+
+Real
+SimpleTestShapeElementKernel::computeQpResidual()
+{
+  return _test[_i][_qp] * _shp_integral;
+}
+
+Real
+SimpleTestShapeElementKernel::computeQpJacobian()
+{
+  return _test[_i][_qp] * _shp_jacobian[_var.dofIndices()[_j]];
```

Comment:
Clever... I like it... maybe we can make it even in a bit better though by hiding some of the indirection...